pkg/reconciler/serverlessservice/serverlessservice.go

```diff
@@ -124,6 +124,59 @@ func (r *reconciler) reconcilePublicService(ctx context.Context, sks *netv1alpha
 	return nil
 }
 
+// subsetEndpoints computes a subset of all endpoints of size `n` using a consistent
+// selection algorithm. For non empty input, subsetEndpoints returns a copy of the
+// input with the irrelevant endpoints and empty subsets filtered out, if the input
+// size is larger than `n`,
+// Otherwise the input is returned as is.
+func subsetEndpoints(eps *corev1.Endpoints, target string, n int) *corev1.Endpoints {
+	if len(eps.Subsets) == 0 {
+		return eps
+	}
+
+	addrs := make([]string, 0, 1)
+	for _, ss := range eps.Subsets {
+		for _, addr := range ss.Addresses {
+			addrs = append(addrs, addr.IP)
+		}
+	}
+
+	// The input is not larger than desired.
+	if len(addrs) <= n {
+		return eps
+	}
+
+	selection := chooseSubset(addrs, n, target)
+
+	// Copy the informer's copy, so we can filter it out.
+	neps := eps.DeepCopy()
+	// Standard in place filter using read and write indices.
+	// This preserves the original object order.
+	r, w := 0, 0
+	for r < len(neps.Subsets) {
+		ss := neps.Subsets[r]
+		// And same algorithm internally.
+		ra, wa := 0, 0
+		for ra < len(ss.Addresses) {
+			if selection.Has(ss.Addresses[ra].IP) {
+				ss.Addresses[wa] = ss.Addresses[ra]
+				wa++
+			}
+			ra++
+		}
+		// At least one address from the subset was preserved, so keep it.
+		if wa > 0 {
+			ss.Addresses = ss.Addresses[:wa]
+			// At least one address from the subset was preserved, so keep it.
+			neps.Subsets[w] = ss
+			w++
+		}
+		r++
+	}
+	neps.Subsets = neps.Subsets[:w]
```

Comment:
IIUC, w should > 0 if n > 0. That's why we don't check the case that w == 1 here?

Reply:
Basically it is based upon a few things
1. a subset cannot be empty, k8s does not permit that.
2. if there are no subsets at all (which is theoretically possible, though practically not), we'd bail out up above `if len(addrs) < n`
3. Thus we can infer that there are at least `n` elements in the response, which means there is at least 1 non empty subset, which means `w > 0`.

Reply:
Added a comment.